ygen/enumgen_test.go

```diff
@@ -20,9 +20,99 @@ import (
 
 	"github.com/google/go-cmp/cmp"
 	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/openconfig/gnmi/errdiff"
 	"github.com/openconfig/goyang/pkg/yang"
 )
 
+func TestResolveDedupSet(t *testing.T) {
+	tests := []struct {
+		name               string
+		inDefinedEnums     map[string]bool
+		inDedupSets        map[string]map[string]bool
+		wantUniqueNamesMap map[string]string
+	}{{
+		name: "no duplication",
+		inDefinedEnums: map[string]bool{
+			"Baz": true,
+		},
+		inDedupSets: map[string]map[string]bool{
+			"Foo": map[string]bool{
+				"enum-a": true,
+			},
+			"Bar": map[string]bool{
+				"enum-b": true,
+			},
+		},
+		wantUniqueNamesMap: map[string]string{
+			"enum-a": "Foo",
+			"enum-b": "Bar",
+		},
+	}, {
+		name: "simple duplication",
+		inDefinedEnums: map[string]bool{
+			"Baz": true,
+		},
+		inDedupSets: map[string]map[string]bool{
+			"Foo": map[string]bool{
+				"enum-a": true,
+				"enum-b": true,
+			},
+		},
+		wantUniqueNamesMap: map[string]string{
+			"enum-a": "Foo",
+			"enum-b": "Foo_",
+		},
+	}, {
+		name: "duplication from existing name",
+		inDefinedEnums: map[string]bool{
+			"Foo": true,
+		},
+		inDedupSets: map[string]map[string]bool{
+			"Foo": map[string]bool{
+				"enum-a": true,
+				"enum-b": true,
+			},
+		},
+		wantUniqueNamesMap: map[string]string{
+			"enum-a": "Foo_",
+			"enum-b": "Foo__",
+		},
+	}, {
+		name: "intersecting dedup sets",
+		inDefinedEnums: map[string]bool{
+			"Baz": true,
+		},
+		inDedupSets: map[string]map[string]bool{
+			// De-dup is done in lexicographical order.
+			"Bar": map[string]bool{
+				"enum-a": true,
+			},
+			"Foo": map[string]bool{
+				// This entry is ignored due to a name being already assigned.
+				"enum-a": true,
```

Comment:
Isn't this an error? `enum-a` can't be mapped to two different types, right?

Reply:
Added `uniqueEnumeratedTypedefEntries map[string]bool{}` and `uniqueEnumeratedLeafEntries map[string]bool{}` to ensure each unique enum can only be mapped to a single default name.